packages/mocha/README.md

```diff
@@ -1 +1,102 @@
+in an application) your tests need to run at just the right moment so
+that they are executed in the correct context. If your tests run too
+soon, they will fail; if they run too slow, well then you just have
+slow tests.
+
+Converging on a state simple means asserting against said state until
+the assertion passes, in which case you have successfully converged on
+that state!
+
+This package uses
+[`@bigtest/convergence`](https://github.com/thefrontside/bigtest/tree/master/packages/convergence)
+to repeatedly run assertions until they pass, or until the timeout has
+almost expired. Performing tests in this way allows them to pass the
+moment the desired state is acheived. This results in very fast tests
+when testing asyncronous things.
+
+_Read the `@bigtest/convergence` [docs on
+convergences](https://github.com/thefrontside/bigtest/tree/master/packages/convergence#why-convergence)
+for more info as to why converging on a desired state is better than
+trying to time it properly._
+
+## Writing Tests
+
+Because convergent assertions are ran repeatedly until they pass, it
+is _highly recommended_ that you **do not** perform any side-effects
+in your assertions. This will result in the side-effect being ran
+multiple times in the case when the assertion fails.
+
+For this reason, you should keep your side-effect producing code in
+hooks and **out of your assertions.** These "pure assertions" also
+help your tests be more readable and explicit.
+
+#### :no_entry_sign: do not do this:
+
+``` javascript
+describe('my button', () => {
+  it('shows a loading indicator after clicking', () => {
+    // this will be called every time the assertion runs
+    $button.click();
+
+    // it might take a few milliseconds for any side-effects to
+    // happen, so this might fail the first time and cause this entire
+    // assertion to run again, thus clicking the button again
+    expect($button.className).to.include('is-loading');
+  });
+});
+```
+
+#### :white_check_mark: do this:
+
+``` javascript
+describe('clicking my button', () => {
+  // keep side-effects inside hooks
+  beforeEach(() => $button.click());
+
+  // a pure assertion has no side-effects; even if it fails, it can
+  // be run again and again without consequence
+  it('shows a loading indicator', () => {
+    expect($button.className).to.include('is-loading');
+  });
+});
+```
+
+### Convergent Hooks?
+
+Sometimes you may attempt to perform an async task to find it fails
+due to a preconcieved state not being met. For example, you can't
+click a button if it doesn't exist in the DOM. You may use
+`@bigtest/convergence` to converge on these states and return a
+promise in your hooks to wait for async tasks.
+
+``` javascript
+describe('clicking my button', () => {
+  // .run() returns a promise that will cause Mocha to wait for the
+  // convergence to resolve before continuing with the assertions
+  beforeEach(() => new Convergence()
```

Comment:
This has me thinking, should we also wrap `beforeEach` to provide `this.convergence` which could be created with the current mocha timeout?

We could also make it so if you return a `Convergence` it will automatically call `.run()`.

Reply:
I like the idea of getting your `run()` for free. Is there a way to detect the timeout of the current running test case globally so that you could just:

```js
import { Convergence } from '@bigtest/mocha'
```

And that constructor would detect the currently running testcase and use its timeout to delegate to the constructor of `@bigtest/convergence`

Reply:
I don't think there is a way to get the current running test's timeout globally. Mocha has `mocha.timeout()` but it's use is for setting the _suite's_ timeout.

The only way I could think of would to update a global `Convergence.timeout` (or some other global) from `it`, which would then be delegated to the constructor of `@bigtest/convergence`.

Reply:
It just occured to me that we could also just create a new convergence off of the old one, that uses the current timeout.

```js
function beforeEach(setup) {
  return mocha.beforeEach(() => {
    let result = setup.apply(this, arguments);
    if (result instanceof Convergence) {
      return result.timeout(this.timeout).run();
    } else {
      return result;
    }
  });  
}